src/client/Microsoft.Identity.Client/ManagedIdentity/AzureArcManagedIdentitySource.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    internal class AzureArcManagedIdentitySource : ManagedIdentitySource
+    {
+        private const string ArcApiVersion = "2019-11-01";
+
+        private readonly string _clientId;
+        private readonly string _resourceId;
+        private readonly Uri _endpoint;
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string imdsEndpoint = EnvironmentVariables.ImdsEndpoint;
+
+            // if BOTH the env vars IDENTITY_ENDPOINT and IMDS_ENDPOINT are set the MsiType is Azure Arc
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(imdsEndpoint))
+            {
+                return default;
+            }
+
+            if (!Uri.TryCreate(identityEndpoint, UriKind.Absolute, out Uri endpointUri))
+            {
+                throw new MsalClientException(MsalError.InvalidManagedIdentityEndpoint, MsalErrorMessage.ManagedIdentityEndpointInvalidUriError);
+            }
+
+            return new AzureArcManagedIdentitySource(endpointUri, requestContext);
+        }
+
+        private AzureArcManagedIdentitySource(Uri endpoint, RequestContext requestContext) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            _clientId = requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedClientId;
+            _resourceId = requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedResourceId;
+
+            if (!string.IsNullOrEmpty(_clientId) || !string.IsNullOrEmpty(_resourceId))
+            {
+                throw new MsalClientException(MsalError.UserAssignedManagedIdentityNotSupported, MsalErrorMessage.ManagedIdentityUserAssignedNotSupported);
```

Comment:
Can we check this once somewhere upstream and not in every provider? Providers can just use clientId if it is set and ignore resourceId

Reply:
This exception is thrown because azure arc does not support user assigned managed identity. This is not supported only for azure arc and cloud shell. The other check that azure sdk have is if both the client id and resource id is passed. We do not have it since we get the values in a single variable.